In [1]:
#importing the library
import pandas as pd

In [2]:
# reading the data from csv and seting the index column
df = pd.read_csv('../data/survey_results_public.csv', index_col='Respondent')
schema_df = pd.read_csv('../data/survey_results_schema.csv', index_col='Column')

In [5]:
# setting the maximum display for rows and column to 85
pd.set_option('display.max_columns', 85)
pd.set_option('display.max_rows', 85)

În cadrul acestei părți din tutorial o să ne uităm peste modalitățile de a grupa sau de a agrega detele dintr-un data frame. Aceste două acțiuni este primul lucru la care se gândesc majoritatea persoanelor când se zice că analizăm date. Aceasta este partea din tutorial unde chiar o să preluăm ceva statistici din data frame, și nu doar modificăm datele din cadrul acestui data frame. De exemplu, poate dorim să știm care este salariul mediu pentru un anumit developer, poate dorim să știm câte persoane din fiecare țară știu Python ca și limbaj de programare. În cadrul acestei părți o să putem răspunde la toate aceste întrebări.

Înainte de a trece la partea mai complicată o să începem cu ceva simplu și o să tot avansăm pe parcursul acestui tutorial. Pentru început o să ne uităm la partea de agregare. Agregarea practic înseamnă combinarea mai multor date într-un singur rezultat. O întrebare la care putem răspunde este care este salariul mediu al unui programator din acest studiu. Pentru asta putem să extragem valoarea medie din coloana unde sunt datele despre salariu. Pentru început o să ne uităm la datele din acea coloană

In [6]:
df['ConvertedComp']

Respondent
1            NaN
2            NaN
3         8820.0
4        61000.0
5            NaN
          ...   
88377        NaN
88601        NaN
88802        NaN
88816        NaN
88863        NaN
Name: ConvertedComp, Length: 88883, dtype: float64

Valorile care sunt cu 'NaN' înseamnă că persoanele care au participat la studiu nu au răspuns la acea întrebare. Pentru a vedea valoarea medie a acestei coloane putem să utilizăm metoda `median()` pentru coloana de 'ConvertedComp'

In [7]:
df['ConvertedComp'].median()

57287.0

Informația de sus este destul de utiliă, însă fiind așa multe persoane din țări diferite, salariile sunt diferite în anumite țări deoarece depinde de la țară la țară ce salarii se oferă, depinde de costul de trai în acea țară și de multe altele. O informație mai utilă ar fi să vedem salariile medii ale participaților per țară (o să ne uităm la asta în partea de grupare a datelor). În continuare o să trecem peste alte câteva concepte basic. Un alt exemplu peste care putem să ne uităm ar fi acela de a aplica anumite metode de agregare pentru întreg data frame-ul, nu doar pentru un obiect de tip Series. O să începem prin a utiliza metoda 'median()' pentru întreg data frame-ul

In [8]:
df.median()

/tmp/ipykernel_666888/530051474.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.median()


CompTotal        62000.0
ConvertedComp    57287.0
WorkWeekHrs         40.0
CodeRevHrs           4.0
Age                 29.0
dtype: float64

Când se utilizează metoda 'median' pentru întreg data frame-ul aceasta o să caute toate coloanele unde sunt date numerice și o să facă media pentru fiecare coloană în parte. Neutilizarea argumentului de 'numeric_only=True' este depricată și o să rezulte cu o eraore pe viitor

In [10]:
df.median(numeric_only=True)

CompTotal        62000.0
ConvertedComp    57287.0
WorkWeekHrs         40.0
CodeRevHrs           4.0
Age                 29.0
dtype: float64

Dacă dorim să vedem o statistică largă a data frame-ului putem utiliza metoda `describe()`. Această metodă o să se aplice din nou doar pentru coloanele care au valori numerice

In [11]:
df.describe()

CompTotal  ConvertedComp   WorkWeekHrs    CodeRevHrs           Age
count  5.594500e+04   5.582300e+04  64503.000000  49790.000000  79210.000000
mean   5.519014e+11   1.271107e+05     42.127197      5.084308     30.336699
std    7.331926e+13   2.841523e+05     37.287610      5.513931      9.178390
min    0.000000e+00   0.000000e+00      1.000000      0.000000      1.000000
25%    2.000000e+04   2.577750e+04     40.000000      2.000000     24.000000
50%    6.200000e+04   5.728700e+04     40.000000      4.000000     29.000000
75%    1.200000e+05   1.000000e+05     44.750000      6.000000     35.000000
max    1.000000e+16   2.000000e+06   4850.000000     99.000000     99.000000

Metoda aceasta de 'describe()' ne afișează mai multe valori despre coloanele ce conțin numere, cum ar fi valoarea minimă, maximă, media și altele. Diferența dintre 'mean()' și 'median()' este că metoda 'mean()' este puternic influențată de către outliers, și anume de către valori care nu sunt concludente. Știm că sunt unele cazuri în care salariul era trecut la 2.000.000 de dolari pe an, iar aceste valori fac ca metoda 'mean()' pentru salarii să returneze valoarea de 1.271107e+05, care practic înseamnă 127.000 și știm că valorea returnată pentru metoda 'meadina()' a fost de 57.000, valoare care este mult mai apropiată de adevăr.

Detaliile despre 'count' rezultate din apicarea metodei 'describe' numără totalul de persoane care au răspuns la acea întrebare. Cele care nu au răspuns sunt trecute cu valoarea de 'NaN', prin urmare această metodă returnează numărul total de rânduri din data frame minus numărul de persoane care au valoarea 'NaN' pentru o anumită coloană. O să utilizăm metoda `count()` separat pentru coloana 'ConvertedComp'

In [12]:
df['ConvertedComp'].count()

55823

Știind că în data frame avem undeva la aproape 89.000 de răspunsuri, metoda de mai sus ne spune că doar undeva la 55.000 de persoane au răspuns la acea întrebare.

O să trecem acuma din nou la coloana 'Hobbyist'. În cadrul acelei coloane persoanele participante la studiu au răspuns cu 'Yes' sau 'No' dacă scriu cod de plăcere sau nu. Dacă dorim să vedem câte persoane au răspuns cu 'Yes' și câte cu 'No' la acea întrebare, atunci putem să utilizăm metoda `value_counts()`

In [13]:
df['Hobbyist'].value_counts()

Yes    71257
No     17626
Name: Hobbyist, dtype: int64

În cadrul acestui studiu există o coloană care deține informații despre ce site de socializare utilizează cei care au răspuns la studiu cel mai mult. Coloana respectivă poartă denumirea de 'SocialMedia'. Pentru a vedea popularitatea acestor site-uri de socializare putem utiliza metoda 'value_counts()'. Pentru început să vedem însă cum arată datele din acea coloană

In [14]:
df['SocialMedia']

Respondent
1          Twitter
2        Instagram
3           Reddit
4           Reddit
5         Facebook
           ...    
88377      YouTube
88601          NaN
88802          NaN
88816          NaN
88863     WhatsApp
Name: SocialMedia, Length: 88883, dtype: object

In [15]:
df['SocialMedia'].value_counts()

Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: SocialMedia, dtype: int64

În situația în care dorim să vedem aceste date sub formă de procente și nu de valori, atunci putem să utilizăm un argument în cadrul metodei 'value_counts()', argument ce poartă denumirea de 'normalize='. Acestui argument o să îi atribuim valorea True

In [17]:
df['SocialMedia'].value_counts(normalize=True)

Reddit                      0.170233
YouTube                     0.163791
WhatsApp                    0.158071
Facebook                    0.156069
Twitter                     0.134988
Instagram                   0.074150
I don't use social media    0.065777
LinkedIn                    0.053306
WeChat 微信                   0.007899
Snapchat                    0.007437
VK ВКонта́кте               0.007141
Weibo 新浪微博                  0.000663
Youku Tudou 优酷              0.000249
Hello                       0.000225
Name: SocialMedia, dtype: float64

Din aceste date, extragem informația cum că 17% dintre respondenți utilizează cel mai des Reddit, 16% YouTube și așa mai departe. În cadrul acelui output putem vedea că există și anumite site-uri de socializare care par a fi doar din anumite zone, cum ar fi China sau Rusia. Întrebarea este cum putem vedea care este cel mai popular site de socializare pentru fiecare țară în parte. Pentru a răspunde la acestă întrebare trebuie să utilizăm conceptul de grupare. Acest concept poate fi un pic confuz pe la început, așa că o să începem cu ce e mai simplu

Dacă dorim să aflăm anumite date bazate pe țări, atunci trebuie să grupăm datele bazate pe coloana 'Country'. Pentru a grupa un set de date avem la dispoziție metoda `groupby()`. Ce anume înseamnă să utilizăm această metodă? În cadrul documentației Pandas, gruparea înseamnă combinarea de împărțire a unui obiect, aplicarea unei funcții și combinare rezultatului. O să trecem prin fiecare proces a acestei părți pentru a înțelege cum anume funcționează.

Pentru început o să ne uităm la numărul de răspunsuri individuale pentru coloana 'Country' ca să ne facem o idee despre ce date avem

In [18]:
df['Country'].value_counts()

United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: Country, Length: 179, dtype: int64

Acum o să ne uităm peste procesul de grupare care presupune cele trei etape de splituire, aplicarea unei funcții și combinarea rezultatelor. O să utilizăm metoda 'groupby()' pentru data frame-ul nostru și o să îi oferim ca și argument numele coloanei (sau coloanelor pentru situațiile în care dorim să grupăm după mai multe coloane) ca și o listă de valori.

In [19]:
df.groupby(['Country'])

Dacă rulăm codul de mai sus ceea ce primim înapoi este un obiect de tipul 'DataFrameGroupBy'. Acest obiect conține o multitudine de grupuri, iar pentru a înțelege mai bine o să ne uităm peste un grup individual al acelui obiect. Pentru a lucra mai ușor cu acel obiect o să îl salvăm într-o variabilă

In [20]:
groupby_country_df = df.groupby(['Country'])

Din moment ce am grupat aceste valori pe baza coloanei 'Country' putem să preluăm un grup din acest obiect pe baza unui nume de țară prezent în data frame. Pentru a prelua un grup din cadrul acestui obiect o să utilizăm metoda `get_group()` la care o să îi atribuim ca și argument o valoare ce reprezintă un nume de țară.

In [21]:
groupby_country_df.get_group('United States')

MainBranch Hobbyist  \
Respondent                                            
4           I am a developer by profession       No   
13          I am a developer by profession      Yes   
22          I am a developer by profession      Yes   
23          I am a developer by profession      Yes   
26          I am a developer by profession      Yes   
...                                    ...      ...   
78292                                  NaN       No   
82717                                  NaN       No   
83397                                  NaN      Yes   
85642                                  NaN       No   
88282                                  NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
4                                                       Never   
13          Less than once a month but more than once per ...   
22                                    Less than once per year   
23                                    Less than once per year   
26                                    Less than once per year   
...                                                       ...   
78292                              Once a month or more often   
82717                                 Less than once per year   
83397                                 Less than once per year   
85642                                 Less than once per year   
88282                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
4           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
22          OSS is, on average, of HIGHER quality than pro...   
23          The quality of OSS and closed source software ...   
26          The quality of OSS and closed source software ...   
...                                                       ...   
78292       OSS is, on average, of HIGHER quality than pro...   
82717       The quality of OSS and closed source software ...   
83397                                                     NaN   
85642       OSS is, on average, of LOWER quality than prop...   
88282       The quality of OSS and closed source software ...   

                                                   Employment        Country  \
Respondent                                                                     
4                                          Employed full-time  United States   
13                                         Employed full-time  United States   
22                                         Employed full-time  United States   
23                                         Employed full-time  United States   
26                                         Employed full-time  United States   
...                                                       ...            ...   
78292       Independent contractor, freelancer, or self-em...  United States   
82717                      Not employed, but looking for work  United States   
83397                      Not employed, but looking for work  United States   
85642       Independent contractor, freelancer, or self-em...  United States   
88282                      Not employed, but looking for work  United States   

           Student                                            EdLevel  \
Respondent                                                              
4               No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
13              No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22              No  Some college/university study without earning ...   
23              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
26              No  Some college/university study without earning ...   
...            ...                                                ...   

Dacă rulăm codul și preluăm acel grup pentru valoarea 'United Stated' putem observa că ni se afișează un data frame. Dacă ne uităm la acel data frame ăn cadrul coloanei 'Country' putem observa că toate valorile sunt 'United States'. Asta conține acest obiect de tipul 'DataFrameGroupBy', a împărțit data frame-ul în grupe în funcție de numele țării. Putem spune că este similar precum rularea unui filtru pe întreg-ul data frame. Ar trebuie să obține același rezultat pentru fiecare țară în parte dacă am crea un filtru și l-am aplica

In [32]:
filter_df = df['Country'] == 'United States'

In [33]:
df.loc[filter_df]

MainBranch Hobbyist  \
Respondent                                            
4           I am a developer by profession       No   
13          I am a developer by profession      Yes   
22          I am a developer by profession      Yes   
23          I am a developer by profession      Yes   
26          I am a developer by profession      Yes   
...                                    ...      ...   
78292                                  NaN       No   
82717                                  NaN       No   
83397                                  NaN      Yes   
85642                                  NaN       No   
88282                                  NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
4                                                       Never   
13          Less than once a month but more than once per ...   
22                                    Less than once per year   
23                                    Less than once per year   
26                                    Less than once per year   
...                                                       ...   
78292                              Once a month or more often   
82717                                 Less than once per year   
83397                                 Less than once per year   
85642                                 Less than once per year   
88282                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
4           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
22          OSS is, on average, of HIGHER quality than pro...   
23          The quality of OSS and closed source software ...   
26          The quality of OSS and closed source software ...   
...                                                       ...   
78292       OSS is, on average, of HIGHER quality than pro...   
82717       The quality of OSS and closed source software ...   
83397                                                     NaN   
85642       OSS is, on average, of LOWER quality than prop...   
88282       The quality of OSS and closed source software ...   

                                                   Employment        Country  \
Respondent                                                                     
4                                          Employed full-time  United States   
13                                         Employed full-time  United States   
22                                         Employed full-time  United States   
23                                         Employed full-time  United States   
26                                         Employed full-time  United States   
...                                                       ...            ...   
78292       Independent contractor, freelancer, or self-em...  United States   
82717                      Not employed, but looking for work  United States   
83397                      Not employed, but looking for work  United States   
85642       Independent contractor, freelancer, or self-em...  United States   
88282                      Not employed, but looking for work  United States   

           Student                                            EdLevel  \
Respondent                                                              
4               No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
13              No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22              No  Some college/university study without earning ...   
23              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
26              No  Some college/university study without earning ...   
...            ...                                                ...   

Dacă dorim să aflăm informații doar despre o anumită țară, atunci procedeul de grupare este foarte asemănător cu cel de filtrare din moment ce retunează aceleași date. În loc să ne returneze rezultatele pentru o singură țară, metoda 'groupby()' splituiește toate răspunsurile după numele țării. Acesta reprezintă primul pas din partea de grupare, și anume splituriea. Din moment ce avem aceste răspunsuri împărțite acuma putem să trecem la pasul doi, acela de a aplica o funcție.

Dacă dorim de exemplu să aflăm cel mai popular site de socializare din Statele Unite, pentru început o să face asta prin utilizarea filtrului de mai sus. Pentur filtrul respectiv o să apelăm metoda 'value_counts()'

In [35]:
df.loc[filter_df]['SocialMedia'].value_counts(normalize=True)

Reddit                      0.284346
Twitter                     0.173002
Facebook                    0.141874
YouTube                     0.122867
I don't use social media    0.092338
Instagram                   0.082410
LinkedIn                    0.050883
WhatsApp                    0.030380
Snapchat                    0.016263
WeChat 微信                   0.004639
VK ВКонта́кте               0.000449
Weibo 新浪微博                  0.000399
Hello                       0.000100
Youku Tudou 优酷              0.000050
Name: SocialMedia, dtype: float64

Acestea sunt valorile doar pentru o țară anume. Dacă ar fi să utilizăm metoda 'value_counts()' pentru obiectul de tip 'DataFrameGroupBy', atunci o să ne calculeze aceste valori pentru fiecare grup în parte. Acel obiect are toate coloanele prezente din data frame-ul mare. Dacă dorim să vizualizăm datele dintr-o coloană, atunci o să ne afișeze doar faptul că este un obiect de tip 'SeriesGroupBy'

In [36]:
groupby_country_df['SocialMedia']

Pentru a vedea toate datele pentru fiecare grup în parte despre coloana 'SocialMedia', atunci pentru acea coloană putem să aplicăm metoda 'value_counts()'

In [37]:
groupby_country_df['SocialMedia'].value_counts()

Country      SocialMedia             
Afghanistan  Facebook                    15
             YouTube                      9
             I don't use social media     6
             WhatsApp                     4
             Instagram                    1
                                         ..
Zimbabwe     Facebook                     3
             YouTube                      3
             Instagram                    2
             LinkedIn                     2
             Reddit                       1
Name: SocialMedia, Length: 1220, dtype: int64

Din moment ce am aplicat acea metodă pentru grupul de date, acesta returnează un obiect de tip Series în care sunt prezente individual, pentru fiecare țară în parte numărul de răspunsuri pentru fiecare valoare din coloana 'SocialMedia'

In [38]:
groupby_country_df['SocialMedia'].value_counts().head(50)

Country              SocialMedia             
Afghanistan          Facebook                     15
                     YouTube                       9
                     I don't use social media      6
                     WhatsApp                      4
                     Instagram                     1
                     LinkedIn                      1
                     Twitter                       1
Albania              WhatsApp                     18
                     Facebook                     16
                     Instagram                    13
                     YouTube                      10
                     Twitter                       8
                     LinkedIn                      7
                     Reddit                        6
                     I don't use social media      4
                     Snapchat                      1
                     WeChat 微信                     1
Algeria              YouTube                      42


După cum spuneam, ceea ce este returnat este un obiect de tip Series, obiect care are mai mulți indexi, și anume 'Country' repsectiv 'SocialMedia'. Primul index este cel pentru 'Country', iar fiind un index putem să preluăm acest index utilizând indexatorul 'loc'. Acest indexator trebuie apleat după ce se rulează metoda 'value_counts()'

In [39]:
groupby_country_df['SocialMedia'].value_counts().loc['United States']

SocialMedia
Reddit                      5700
Twitter                     3468
Facebook                    2844
YouTube                     2463
I don't use social media    1851
Instagram                   1652
LinkedIn                    1020
WhatsApp                     609
Snapchat                     326
WeChat 微信                     93
VK ВКонта́кте                  9
Weibo 新浪微博                     8
Hello                          2
Youku Tudou 优酷                 1
Name: SocialMedia, dtype: int64

In [40]:
groupby_country_df['SocialMedia'].value_counts().loc['India']

SocialMedia
WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: SocialMedia, dtype: int64

Dacă dorim să aflăm date despre anumite țări rapid, este destul să schimbăm numele index-ului pe care dorim să îl extragem. De aceasta este destul de folositoare partea de grupare a datelor

În acest moment putem să utilizăm metode de agregare peste aceste grupuri (precum 'median()', 'mean()' ). Pentru a vedea salariile medii din fiecare țară prezentă în studiu trebuie să accesăm coloana 'ConvertedComp' iar pentru aceasta să rulăm metoda 'median()'

In [41]:
groupby_country_df['ConvertedComp'].median()

Country
Afghanistan                               6222.0
Albania                                  10818.0
Algeria                                   7878.0
Andorra                                 160931.0
Angola                                    7764.0
                                          ...   
Venezuela, Bolivarian Republic of...      6384.0
Viet Nam                                 11892.0
Yemen                                    11940.0
Zambia                                    5040.0
Zimbabwe                                 19200.0
Name: ConvertedComp, Length: 179, dtype: float64

Rezultatul este un obiect de tip Series unde index-ul reprezintă un nume de țară. Dacă dorim să aflăm salariul mediu dintr-o anumită țară putem să utilizăm indexatorul 'loc' pentru a extrage acel index

In [42]:
groupby_country_df['ConvertedComp'].median().loc['Germany']

63016.0

Să trecem acuma peste cazul în care lucrăm cu ceva anilze și dorim să grupăm datele, dar dorim și să rulăm mai multe date de agregare pentru grupurile respective. Să luăm cazul în care dorim să apelăm atât metoda 'median()' cât și cea de 'mean()'. Pentru a face acest lucru putem să ne folosim de metoda `agg()` la care să îi pasăm un număr de metode ce dorim să le aplicăm. Metodele trebuie specificate într-o listă sub formă de string

In [43]:
groupby_country_df['ConvertedComp'].agg(['median', 'mean'])

median           mean
Country                                                      
Afghanistan                             6222.0  101953.333333
Albania                                10818.0   21833.700000
Algeria                                 7878.0   34924.047619
Andorra                               160931.0  160931.000000
Angola                                  7764.0    7764.000000
...                                        ...            ...
Venezuela, Bolivarian Republic of...    6384.0   14581.627907
Viet Nam                               11892.0   17233.436782
Yemen                                  11940.0   16909.166667
Zambia                                  5040.0   10075.375000
Zimbabwe                               19200.0   34046.666667

[179 rows x 2 columns]

În funcție de ce anume dorim să realizăm, e posibil să ne batem de ceva erori care nu ne-am aștepta să se întâmple. Să luăm cazul în care încercăm să ne dăm seama câte persoane din fiecare țară știu Python ca limbaj de programare. Înainte de a ne uita cum putem face asta pentru un grup, o să realizăm acest pas pentru o singură coloană utilizând partea de filtrare.

In [44]:
filter_df = df['Country'] == 'India'

In [45]:
df.loc[filter_df]

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   
...                                            ...      ...   
77339                                          NaN      Yes   
79795                                          NaN      Yes   
83862                                          NaN      Yes   
84299                                          NaN      Yes   
86012                                          NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
8                                     Less than once per year   
10                                 Once a month or more often   
15                                                      Never   
50                                 Once a month or more often   
65                                                      Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   
...                                        ...     ...             ...   
77339                       Employed full-time   India  Yes, full-time   
79795                       Employed full-time   India              No   
83862                       Employed full-time   India  Yes, full-time   
84299                       Employed full-time   India  Yes, full-time   
86012                       Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                       ...   
77339                Bachelor’s degree (BA, B

Pentru a vedea câte persoane știu Python o să ne folosim de metodele specifice pentru string-uri pe care le-am mai tot folosit pe parcursul tutorialului. Datele respective sunt prezente în cadrul coloanei 'LanguageWorkedWith'. Din cadrul acestui data frame rezultat după filtrare putem să extragem doar acea coloană sub formă se Series.

In [49]:
df.loc[filter_df]['LanguageWorkedWith']

Respondent
8        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
10                           C#;Go;JavaScript;Python;R;SQL
15       Assembly;Bash/Shell/PowerShell;C;C++;HTML/CSS;...
50       Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
65                         Assembly;C;C++;C#;HTML/CSS;Java
                               ...                        
77339                                            Other(s):
79795                 Bash/Shell/PowerShell;Python;SQL;VBA
83862    Assembly;C;C++;HTML/CSS;Java;JavaScript;Object...
84299     HTML/CSS;Java;JavaScript;Python;Swift;TypeScript
86012        Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript
Name: LanguageWorkedWith, Length: 9061, dtype: object

Acum, asupra acelui obiect de tip Series putem utiliza clasa 'str', iar din acea clasă să utilizăm metoda 'containes()' pentru a verifica dacă acele string-uri conțin Python sau nu

In [48]:
df.loc[filter_df]['LanguageWorkedWith'].str.contains('Python')

Respondent
8         True
10        True
15       False
50        True
65       False
         ...  
77339    False
79795     True
83862    False
84299     True
86012    False
Name: LanguageWorkedWith, Length: 9061, dtype: object

Pentru a putea verifica numărul total de persoane care știu Python putem utiliza metoda `sum()`. Poate ne-am imagina că acea metodă funcționează doar pentru valori integer sau float, dar funcționează și pentru valori de tipul boolean. Metoda o putem apela pentru codul de mai sus care ne afișează acele valori booleane

In [50]:
df.loc[filter_df]['LanguageWorkedWith'].str.contains('Python').sum()

3105

Înainte, în momentul în care doream să apelăm o funcție de agregare similară pentru obiectul de tip 'DataFrameGroupBy' pur și simplu utilizam același aproach pentru acel obiect. Poate ne gândinm că în loc să utilizăm acel filtru (df.loc[filter_df]) putem să utilizăm obiectul respectiv și să păstrăm restul codului și ar trebuie să obținem rezultatul care dorim.

In [51]:
groupby_country_df['LanguageWorkedWith'].str.contains('Python').sum()

AttributeError: 'SeriesGroupBy' object has no attribute 'str'

Dacă rulăm acel cod, atunci o să ne apară o eroare. Eroarea ne spune că obiectul de tip 'SeriesGroupBy' nu are atributul 'str' implementat. Acest obiect conține un grup de obiecte de tip Series. Pentru a putea rula o funcție pentru fiecare Series din acel obiect trebuie să utilizăm metoda 'apply()' căreia îi spunem ce metodă să apeleze pentru fiecare obiect de tip Series din cadrul grupului. Funcția respectivă poate fi creată înainte sau se poate utiliza o funcție lambda

In [52]:
groupby_country_df['LanguageWorkedWith'].apply(lambda x : x.str.contains('Python').sum())

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

Dacă rulăm acel cod putem observa că obținem numărul persoanelor din fiecare țară care au răspuns că știu ca și limbaj de programare 'Python' la acest studiu. Doar aceste numere nu ne ajută așa de mult dacă am dori să ne facem o idee despre procentajul persoanelor care știu Python din fiecare țară. Pentru a face acest lucru o să ne folosim de mai multe metode ce le-am învățat în cadrul acestui tutorial. Cel mai probabil există mai multe metode de a afla procentajul respectiv

Pentru început o să extragem numărul total de persoane care au participat la studiu din fiecare țară

In [53]:
country_respondents = df['Country'].value_counts()

In [54]:
country_respondents

United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: Country, Length: 179, dtype: int64

Următorul pas este să salvăm datele ce reprezintă numărul persoanelor care știu Python din fiecare țară.

In [55]:
country_respondents_use_python = groupby_country_df['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())

In [56]:
country_respondents_use_python

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

În acest moment avem două obiecte de tip Series, unul ce conține numărul persoanelor din fiecare țară care au răspuns la acest studiu, iar alt Series conține numărul persoanelor din fiecare țară care știu Python. Urmează să combină aceste două Series. Pentru a le combina o să utilizăm o metodă pe care încă nu am utilizat-o până în acest moment. Putem utiliza metoda `concat()` pentru a cread un data frame din mai multe obiecte de tip Series. Această metodă se apelează pentru librăria Pandas în sine, iar ca și argument trebuie să îi oferim o listă care să conțină numele obiectelor de tip Series ce dorim să le concatenăm

In [ ]:
python_df = pd.concat([country_respondents, country_respondents_use_python])

Pentru codul de mai sus mai trebuie să adăugăm încă un argument, și anume axa pe care să se facă această concatenare. În mod inițial această axă este setată la 'rows' și încearcă să facă concatenarea pe rânduri, dar noi vrei să matchuim datele dintr-un Series în altul după index, prin urmare trebuie să schimbăm axa să fie pe coloane. Pentru a schimba această axă se va utiliza argumentul `axis=`

In [57]:
python_df = pd.concat([country_respondents, country_respondents_use_python], axis='columns')

In [58]:
python_df

Country  LanguageWorkedWith
United States        20949               10083
India                 9061                3105
Germany               5866                2451
United Kingdom        5737                2384
Canada                3395                1558
...                    ...                 ...
Tonga                    1                   0
Timor-Leste              1                   1
North Korea              1                   0
Brunei Darussalam        1                   0
Chad                     1                   0

[179 rows x 2 columns]

În acest moment avem un data frame care a fost creat din concatenarea a două obiecte de tip Series, iar concatenarea aceasta s-a produs pe baza index-ului, și anume pe baza numelui țării. În cadrul acestui data frame avem două coloane care nu sunt relevante pentru ce date rețin. O să redenumim aceste coloane pentru a fi cât mai concrete în legătură cu ce date țin.

In [60]:
python_df.rename(columns={'Country' : 'NumRespondents', 'LanguageWorkedWith' : 'NumKnowsPython'}, inplace=True)

In [61]:
python_df

NumRespondents  NumKnowsPython
United States               20949           10083
India                        9061            3105
Germany                      5866            2451
United Kingdom               5737            2384
Canada                       3395            1558
...                           ...             ...
Tonga                           1               0
Timor-Leste                     1               1
North Korea                     1               0
Brunei Darussalam               1               0
Chad                            1               0

[179 rows x 2 columns]

În acest moment avem cele două coloane de interes, mai rămâne numai să creem o nouă coloană cu procentajul pentru fiecare valoare. Pentru a crea acel procentaj o să utilizăm regula de 3 simplă și o să salvăm acel rezultat într-o coloană nouă.

In [62]:
python_df['PercKnowsPython'] = (python_df['NumKnowsPython'] / python_df['NumRespondents']) * 100

In [63]:
python_df

NumRespondents  NumKnowsPython  PercKnowsPython
United States               20949           10083        48.131176
India                        9061            3105        34.267741
Germany                      5866            2451        41.783157
United Kingdom               5737            2384        41.554820
Canada                       3395            1558        45.891016
...                           ...             ...              ...
Tonga                           1               0         0.000000
Timor-Leste                     1               1       100.000000
North Korea                     1               0         0.000000
Brunei Darussalam               1               0         0.000000
Chad                            1               0         0.000000

[179 rows x 3 columns]

În acest moment coloana respectivă din acest data frame se poate utiliza normal ca orice coloană din cadrul unui data frame. De exemplu putem să sortăm aceste valori sau să aplicăm anumite metode de agregare pentru acestea.

In [64]:
python_df.sort_values(by='PercKnowsPython', ascending=False)

NumRespondents  NumKnowsPython  \
Sao Tome and Principe                          1               1   
Timor-Leste                                    1               1   
Dominica                                       1               1   
Niger                                          1               1   
Turkmenistan                                   7               6   
...                                          ...             ...   
Cape Verde                                     3               0   
Lao People's Democratic Republic               3               0   
Malawi                                         2               0   
Liberia                                        2               0   
Chad                                           1               0   

                                  PercKnowsPython  
Sao Tome and Principe                  100.000000  
Timor-Leste                            100.000000  
Dominica                               100.000000  
Niger                                  100.000000  
Turkmenistan                            85.714286  
...                                           ...  
Cape Verde                               0.000000  
Lao People's Democratic Republic         0.000000  
Malawi                                   0.000000  
Liberia                                  0.000000  
Chad                                     0.000000  

[179 rows x 3 columns]

In [65]:
python_df['PercKnowsPython'].median()

34.69387755102041

## Recapitulare

În această parte din tutorial am învățat următoarele lucruri:

    1. utilizarea metodelor de tip agregare pentru o coloană de date (de tip numeric)

        df['ConvertedComp'].median()

    2. Utilizarea metodelor de tip agregare pentru întreg data frame-ul

        df.media(numeric_only=True)

    3. Utilizarea metodei rescribe pentru întreg data frame-ul

        df.describe()

            # returnează date despre întreg data frame-ul (pentru coloanele ce conțin valori numerice), iar aceste date sunt reprezentate de mai multe metode de agregare

    4. Metoda de a număra totalitatea răspunsurilor dintr-o anumită coloană (fără cele care au valoarea 'NaN')

        df['ConvertedComp'].count()

    5. Metoda de a număra fiecare răspuns în parte din cadrul unei coloane

        df['Hobbyist'].value_counts()

            # returnează 'Yes' = 71257, 'No' = 17626 sub formă de Series

    6. Afișarea perocentajului pentru metoda 'value_counts()'

        df['SocialMedia'].value_counts(normalize=True)

    7. Utilizarea metodei groupby

        groupby_country_df = df.groupby(['Country'])

            # returnează un obiect de tip DataFrameGroupBy ce conține mai multe grupuri, unul pentru fiecare țară în parte

    8. Accesarea unui grup dintr-un DdataFrameGroupBy

        groupby_countr_df.get_group('United States')

    9. Utilizarea unei metode pentru un întreg obiect de grupuri

        groupby_country_df['SocialMedia'].value_counts()

    10. Accesarea unui singur grup din cadrul unui obiect de grupuri în urma unei funcții aplicate

        groupby_country_df['SocialMedia'].value_counts().loc['United States']

    11. Utilizarea unei metode de agregare pentru întregul obiect de grupuri

        groupby_country_df['ConvertedComp'].median()

    12. Utilizarea mai multor metode de agregare pentru întreg obiectul de grupuri 

        groupby_country_df['ConvertedComp'].agg(['median', 'mean'])

    13. Aplicarea unei metode pentru un grup de obiecte de tipul Series

        groupby_country_df['LanguageWorkedWith'].apply(lambda x : x.str.contains('Python').sum())

    14. Concatenarea a două obiecte de tip Series într-un data frame

        country_res = df['Country'].value_counts()

        country_res_python = groupby_country_df['LanguageWorkedWith'].apply(lambda x : x.str.contains('Python').sum())

        python_df = pd.concat([country_res, country_res_pytho])

    15. Scimbarea numelor coloanelor din noul data frame 

        python_df.rename(columns={'Country' : 'NumRespondents', 'LanguageWorkedWith' : 'NumKnowsPython'}, inplace=True)

    16. Adăugarea unei noi coloane care reprezintă procentajul persoanelor care știu Python din fiecare țară

        python_df['PercKnowsPython'] = (python_df['NumKnowsPython'] / python_df['NumRespondents']) * 100

            